### ISSUE TITLE
Tree merge API

### ISSUE THREAD

**Issue by joshtriplett** [libgit2/libgit2#769]
I'd like to merge two or more trees into a single tree.  Such a merge follows a conceptually simple algorithm: walk the trees in parallel, include blobs or subtrees only present in one or the other, include blobs or subtrees present in both with identical hashes, recurse into subtrees present in both with different hashes, abort on blobs present in both with different hashes, abort on blob-versus-subtree.  ("abort" could become "run custom handler", but in my case I don't need to handle imperfect merges.)

I could easily enough write most of this algorithm on top of the existing libgit2 API, except for a couple of issues:
- As far as I can tell, when building a tree with a treebuilder, I don't have any way to efficiently walk the treebuilder and a tree to merge into it in parallel, comparing corresponding objects.  I'd have to look up each of the tree's entries in the treebuilder separately, and do a separate insert for each object that doesn't exist (or each tree constructed from a recursive merge operation).
- This seems like a sufficiently common use case to warrant inclusion in libgit2 rather than duplicate code in many projects using libgit2.
